src/mudlet.cpp

```diff
@@ -4122,6 +4122,553 @@ void mudlet::playSound(const QString& s, int soundVolume)
+
+        if (!soundFile.exists() && !mudlet::self()->downloadMSPFile(soundFileName, soundType, fileName)) {
+            return; // Exit if no local file and no downloaded file.
+        }
+    }
+
+    QMediaPlayer* pPlayer = mudlet::self()->getMSPSoundMediaPlayer();
+
+    if (!pPlayer) { // It should be impossible to ever reach this.
+        TDebug(QColor(Qt::white), QColor(Qt::red)) << QStringLiteral("mudlet::playMSPSound() WARNING - Unable to create new QMediaPlayer object\n") >> 0;
+        return;
+    }
+
+    if (soundLength == 1) { // Play once
+        if (fileNameList.size() > 1) {
+            pPlayer->setMedia(QUrl::fromLocalFile(fileNameList.at(qrand() % fileNameList.size())));
+        } else {
+            pPlayer->setMedia(QUrl::fromLocalFile(fileNameList.at(0)));
+        }
+    } else {
+        QMediaPlaylist* playlist = new QMediaPlaylist;
+
+        if (soundLength == -1) { // Repeat indefinitely
+            if (fileNameList.size() > 1) {
+                playlist->addMedia(QUrl::fromLocalFile(fileNameList.at(qrand() % fileNameList.size())));
+            } else {
+                playlist->addMedia(QUrl::fromLocalFile(fileNameList.at(0)));
+            }
+            playlist->setPlaybackMode(QMediaPlaylist::Loop);
+        } else {
+            for (int k = 0; k < soundLength; k++) { // Repeat a finite number of times
+                if (fileNameList.size() > 1) {
+                    playlist->addMedia(QUrl::fromLocalFile(fileNameList.at(qrand() % fileNameList.size())));
+                } else {
+                    playlist->addMedia(QUrl::fromLocalFile(fileNameList.at(0)));
+                }
+            }
+        }
+
+        playlist->setCurrentIndex(1);
+        pPlayer->setPlaylist(playlist);
+    }
+
+    // Set volume and play
+    pPlayer->setVolume(soundVolume);
+    pPlayer->play();
+}
+
+QMediaPlayer* mudlet::getMSPMusicMediaPlayer(int soundLength, int musicContinue, QString absolutePathFileName)
+{
+    QMediaPlayer* pPlayer = nullptr;
+    QPointer<Host> pHost = getActiveHost();
+
+    if (pHost) {
+        if (musicContinue == 1) { // If the same music file is playing, continue playing it instead of restarting
+            QListIterator<QMediaPlayer*> itMusicBox1(mMusicBoxList);
+
+            while (itMusicBox1.hasNext()) {
+                QMediaPlayer* pTestPlayer = itMusicBox1.next();
+
+                if (pTestPlayer->state() == QMediaPlayer::PlayingState && pTestPlayer->mediaStatus() != QMediaPlayer::LoadingMedia) {
+                    if (pTestPlayer->media().canonicalUrl() == absolutePathFileName) {
+                        // If the same music is already playing, and we've not been instructed to repeat, allow it to continue and exit here
+                        if (soundLength == 1) {
+                            return pPlayer;
+                        } else { // Continue playing the music, but add more instances below
+                            pPlayer = pTestPlayer;
```

Comment:
Don't you need a `break;`  here to leave the `while` loop to find a player already playing the sound now we have done so...?

Reply:
I did need a `break;` there, good catch.  I had a `return` there before, which was also a mistake, and I removed it too quickly today and did not replace it with the `break;`.  Resolving.